internal/controller/machineaccount_controller.go

```diff
@@ -0,0 +1,256 @@
+		}
+		log.Info("Machine account already deleted from Zitadel", "username", machineAccount.GetUID())
+	}
+
+	log.Info("Successfully finalized machine account", "username", machineAccount.GetUID())
+	return finalizer.Result{}, nil
+}
+
+// +kubebuilder:rbac:groups=iam.miloapis.com,resources=machineaccounts,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=iam.miloapis.com,resources=machineaccounts/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=iam.miloapis.com,resources=machineaccounts/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the MachineAccount object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.21.0/pkg/reconcile
+func (r *MachineAccountController) Reconcile(ctx context.Context, req mcreconcile.Request) (ctrl.Result, error) {
+	log := logf.FromContext(ctx).WithName("machineaccount-reconciler")
+	log.Info("Starting reconciliation", "request", req)
+
+	machineAccount := &iammiloapiscomv1alpha1.MachineAccount{}
+	err := r.Client.Get(ctx, req.NamespacedName, machineAccount)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("MachineAccount resource not found")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get MachineAccount resource")
+		return ctrl.Result{}, err
+	}
+
+	// Funalizers must run before we update the MachineAccount Status,
+	// we need to set the Status State to the desired state on the first complete reconciliation
+	log.Info("Running finalizers", "machineAccountName", machineAccount.GetName(), "machineAccountUID", machineAccount.GetUID())
+	finalizeResult, err := r.Finalizers.Finalize(ctx, machineAccount)
+	if err != nil {
+		log.Error(err, "Failed to run finalizers for MachineAccount")
+		return ctrl.Result{}, fmt.Errorf("failed to run finalizers for MachineAccount: %w", err)
+	}
+	if finalizeResult.Updated {
+		log.Info("Finalizer updated the machineAccount object, updating API server")
+		if updateErr := r.Client.Update(ctx, machineAccount); updateErr != nil {
+			log.Error(updateErr, "Failed to update MachineAccount after finalizer update")
+			return ctrl.Result{}, updateErr
+		}
+		return ctrl.Result{}, nil
+	}
+
+	maComputedEmail := r.computeEmailAddress(machineAccount, req)
+	log.Info("Checking if machine account exists in Zitadel", "username", machineAccount.GetUID())
+	_, err = r.Zitadel.GetUser(ctx, string(machineAccount.GetUID()))
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Error(err, "Failed to get user from Zitadel", "username", machineAccount.GetUID())
+			return ctrl.Result{}, err
```

Comment:
I'd recommend wrapping these errors when returning them so it's easier to trace back to where the error was emitted. Same feedback applies to other places where an error is returned. 

```suggestion
			return ctrl.Result{}, fmt.Error("failed to get user from Zitadel: %w", err)
```